src/http/httpProxy.ts

```diff
@@ -0,0 +1,119 @@
+// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the MIT License.
+
+import { serialize as serializeCookie } from 'cookie';
+import { EventEmitter } from 'events';
+import * as http from 'http';
+import { AzFuncSystemError, ensureErrorType } from '../errors';
+import { nonNullProp } from '../utils/nonNull';
+import { workerSystemLog } from '../utils/workerSystemLog';
+import { HttpResponse } from './HttpResponse';
+
+const requests: Record<string, http.IncomingMessage> = {};
+const responses: Record<string, http.ServerResponse> = {};
+
+const invocRequestEmitter = new EventEmitter();
+
+export async function waitForRequest(invocId: string): Promise<http.IncomingMessage> {
+    return new Promise((resolve, _reject) => {
+        const req = requests[invocId];
+        if (req) {
+            resolve(req);
+            delete requests[invocId];
+        } else {
+            invocRequestEmitter.once(invocId, () => {
+                const req = requests[invocId];
+                if (req) {
+                    resolve(req);
+                    delete requests[invocId];
+                }
+            });
+        }
+    });
+}
+
+const invocIdHeader = 'x-ms-invocation-id';
+export async function sendResponse(invocId: string, userRes: HttpResponse): Promise<void> {
+    const proxyRes = nonNullProp(responses, invocId);
+    delete responses[invocId];
+    for (const [key, val] of userRes.headers.entries()) {
+        proxyRes.setHeader(key, val);
+    }
+    proxyRes.setHeader(invocIdHeader, invocId);
+    proxyRes.statusCode = userRes.status;
+
+    if (userRes.cookies.length > 0) {
+        setCookies(userRes, proxyRes);
+    }
+
+    if (userRes.body) {
+        for await (const chunk of userRes.body.values()) {
+            proxyRes.write(chunk);
+        }
+    }
+    proxyRes.end();
+}
+
+function setCookies(userRes: HttpResponse, proxyRes: http.ServerResponse): void {
+    const serializedCookies: string[] = userRes.cookies.map((c) => {
+        let sameSite: true | false | 'lax' | 'strict' | 'none' | undefined;
+        switch (c.sameSite) {
+            case 'Lax':
+                sameSite = 'lax';
+                break;
```

Comment:
We could also do something like:

```
c.sameSite = typeof c.sameSite === "string" ? c.sameSite.toLowerCase() : c.sameSite;
```

If we don't want to check against `string`, we could create a union type of lax, strict, and none.

Reply:
If I do that I get a TypeScript error:

> Type 'string | undefined' is not assignable to type 'boolean | "lax" | "strict" | "none" | undefined'.
  Type 'string' is not assignable to type 'boolean | "lax" | "strict" | "none" | undefined'.

The only reason I have to do this is because rpc and the cookie parser both use string constant union types, but unfortunately they handle the casing differently. If I use `toLowerCase`, I lose type checking because it's converted to a generic string instead of the string constant union type